### PR TITLE
ARIA-149 Enhance operation configuration

### PR DESCRIPTION
**Pull request by tliron**
* Parse special "dependencies" configuration parameters as YAML and
  treat as Parameter models, allowing them full use of intrinsic
  functions, type coersions, and validations
* Rename various functions that process "properties" to more generically
  process "parameters" (properties, inputs, attributes, arguments, etc.)
* The "configuration" field in OperationTemplate and Operation models
  is now now a dict of Parameter models
* Added "function" and "arguments" fields to Operation model to preserve
  user data (in "implementation" and "inputs") and to clearly demarcate
  orchestration data from user data
* Some cleanup of parser code touched by this commit

### CODE REVIEW

File: aria/modeling/exceptions.py [hunk 1]
```diff
@@ -22,9 +22,9 @@ class ModelingException(AriaException):
     """
 
 
-class InputsException(ModelingException):
+class ParameterException(ModelingException):
```
Comment (ran-z) [inline aria/modeling/exceptions.py]: I'm not sure I understand this change - It's true that the model object is `Parameter` and not `Inputs` (although that too is supposedly under change by @AviaE ) - but the error itself is not about the model type and doesn't have anything much to do with the models themselves - it's about inputs, so why shouldn't the exception be named like so?

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml [hunk 2]
```diff
@@ -309,7 +311,7 @@ policy_types:
       client connections cleanly and shut down services. 
     derived_from: aria.Workflow
     properties:
-      implementation:
+      function:
```
Comment (ran-z) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: i thought in the service-template even according to your view this should remain `implementation`, no?

File: extensions/aria_extension_tosca/simple_v1_0/modeling/parameters.py [hunk 3]
```diff
@@ -71,6 +72,7 @@ def get_assigned_and_defined_property_values(context, presentation, field_name='
     values = OrderedDict()
 
     the_type = presentation._get_type(context)
+    field_name_plural = pluralize(field_name)
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/parameters.py]: how did the following lines worked before this? o_O

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py [hunk 4]
```diff
@@ -384,18 +387,37 @@ def create_operation_template_model(context, service_template, operation):
...
+                    # Parse as YAML
+                    try:
+                        value = yaml.load(value)
+                    except yaml.parser.MarkedYAMLError as e:
+                        context.validation.report(
+                            'YAML parser {0} in operation configuration: {1}'
+                            .format(e.problem, value),
+                            locator=implementation._locator,
+                            level=Issue.FIELD)
+                        continue
```
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: shouldn't this be a re-raise? seems somewhat fatal...

File: aria/orchestrator/execution_plugin/instantiation.py [hunk 5]
```diff
@@ -16,107 +16,132 @@
...
-    ssh = _get_ssh(configuration.pop('ssh')) if 'ssh' in configuration else {}
     if 'user' not in ssh:
         ssh['user'] = default_user
     if ('password' not in ssh) and ('key' not in ssh) and ('key_filename' not in ssh):
         ssh['password'] = default_password
 
-    arguments['use_sudo'] = ssh.get('use_sudo', False)
-    arguments['hide_output'] = ssh.get('hide_output', [])
-    arguments['fabric_env'] = {}
+    operation.arguments['use_sudo'] = Parameter.wrap('use_sudo', ssh.get('use_sudo', False),
```
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: shouldn't the wrapping of configuration parameters simply take place where this module is called from?
what about configuration parameters not meant for the execution plugin?

File: aria/modeling/service_instance.py [hunk 6]
```diff
@@ -1728,34 +1732,41 @@ def operation_template(cls):
...
-            # can validate the configuration and otherwise return specially derived arguments
-            arguments = self.configuration
+            # can validate the configuration and otherwise create specially derived arguments. For
+            # now, we just send all configuration parameters as arguments
+            utils.instantiate_dict(self, self.arguments, self.configuration)
 
-        # Note: the arguments will *override* operation inputs of the same name
-        if arguments:
-            for k, v in arguments.iteritems():
-                self.inputs[k] = models.Parameter.wrap(k, v)
```
Comment (ran-z) [inline aria/modeling/service_instance.py]: why remove this?

Comment (ran-z) [inline aria/modeling/service_instance.py]: :+1:

File: aria/modeling/exceptions.py (hunk 1 above)
Comment (tliron) [inline aria/modeling/exceptions.py]: Discussed and we agreed that more general is better in this case.

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml (hunk 2 above)
Comment (tliron) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: Correct, changed back.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/parameters.py (hunk 3 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/parameters.py]: Previously we sent the plural as a function argument, but I optimized by removing it... just some code cleanup.

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 4 above)
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: Validation issues accumulate and are displayed in the end. None are considered fatal enough to start the parsing, but of course a single error means that parsing failed.

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 5 above)
Comment (tliron) [inline aria/orchestrator/execution_plugin/instantiation.py]: We went through the code together, this happens elsewhere.

File: aria/modeling/utils.py [hunk 7]
```diff
@@ -52,84 +53,95 @@ def service_template(self):
         return self.container.service_template
 
 
-def create_parameters(parameters, declared_parameters):
+def merge_parameter_values(parameter_values, declared_parameters, forbidden_names=None):
```
Comment (ran-z) [inline aria/modeling/utils.py]: i thought we talked about the concept of validating "forbidden names" and we agreed it should be done in the parser / instantiation phases instead. I don't see in what scenario we'd like to inform the user of such validation errors only at this stage.

File: aria/modeling/exceptions.py [hunk 8]
```diff
@@ -57,3 +57,9 @@ class UndeclaredParametersException(ParameterException):
     """
     ARIA modeling exception: Undeclared parameters have been provided.
     """
+
+
+class ForbiddenParameterNamesException(ParameterException):
```
Comment (ran-z) [inline aria/modeling/exceptions.py]: perhaps "Reserved" is more appropriate than "Forbidden"?

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 5 above)
Comment (ran-z) [inline aria/orchestrator/execution_plugin/instantiation.py]: It's true that configuration parameters are wrapped in another place when it's not intended for the execution plugin, but my question is why bother wrapping them at this stage at all - as opposed to working with dicts at this stage and then have the instantiation code which calls this module do the wrapping. it just makes more sense IMO.

File: aria/modeling/exceptions.py (hunk 1 above)
Comment (ran-z) [inline aria/modeling/exceptions.py]: :+1:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/parameters.py (hunk 3 above)
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/parameters.py]: :+1:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py (hunk 4 above)
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/__init__.py]: :+1:

File: tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml (hunk 2 above)
Comment (ran-z) [inline tests/resources/service-templates/tosca-simple-1.0/node-cellar/node-cellar.yaml]: :+1:

File: tests/orchestrator/test_workflow_runner.py [hunk 9]
```diff
@@ -259,8 +258,9 @@ def _setup_mock_workflow_in_service(request, inputs=None):
     workflow = models.Operation(
         name=mock_workflow_name,
         service=service,
-        implementation='workflow.mock_workflow',
-        inputs=inputs or {})
+        function='workflow.mock_workflow',
+        inputs=inputs or {},
+        arguments=inputs or {})
```
Comment (ran-z) [inline tests/orchestrator/test_workflow_runner.py]: why is this needed? wouldn't arguments get auto-populated with inputs anyway?

File: aria/orchestrator/execution_plugin/instantiation.py (hunk 5 above)
Comment (tliron) [inline aria/orchestrator/execution_plugin/instantiation.py]: If you notice I am also giving the parameters descriptions here. These are not mere dicts, but true parameters.

File: aria/orchestrator/workflow_runner.py [hunk 10]
```diff
@@ -136,10 +137,11 @@ def _validate_no_active_executions(self, execution):
         active_executions = [e for e in self.service.executions if e.is_active()]
         if active_executions:
             raise exceptions.ActiveExecutionsError(
-                "Can't start execution; Service {0} has an active execution with id {1}"
+                "Can't start execution; Service {0} has an active execution with ID {1}"
                 .format(self.service.name, active_executions[0].id))
 
-    def _get_workflow_fn(self):
+    @property
+    def _workflow_fn(self):
```
Comment (tliron) [inline aria/orchestrator/workflow_runner.py]: Changing back to function.

File: aria/modeling/utils.py [hunk 11]
```diff
@@ -51,74 +53,95 @@ def service_template(self):
         return self.container.service_template
 
 
-def create_inputs(inputs, template_inputs):
+def merge_parameter_values(parameter_values, declared_parameters, forbidden_names=None):
```
Comment (ran-z) [inline aria/modeling/utils.py]: what was the purpose of the changes to this method? im not sure i understand the benefit and it seems less readable now to me

File: aria/orchestrator/workflows/events_logging.py [hunk 12]
```diff
@@ -35,20 +35,20 @@ def _get_task_name(task):
...
 def _start_task_handler(task, **kwargs):
-    # If the task has not implementation this is an empty task.
-    if task.implementation:
+    # If the task has no function this is an empty task.
+    if task.function:
         suffix = 'started...'
         logger = task.context.logger.info
     else:
-        suffix = 'has no implementation'
+        suffix = 'has no function'
```
Comment (tliron) [inline aria/orchestrator/workflows/events_logging.py]: Changing back to "implementation".

File: extensions/aria_extension_tosca/simple_v1_0/modeling/functions.py [hunk 13]
```diff
@@ -69,7 +69,7 @@ def __evaluate__(self, container_holder):
             e, final = evaluate(e, final, container_holder)
             if e is not None:
                 value.write(unicode(e))
-        value = value.getvalue()
+        value = value.getvalue() or u''
```
Comment (tliron) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/functions.py]: We default to Unicode everywhere possible in values -- this avoids thorny problems later down the road with values.

File: aria/modeling/exceptions.py (hunk 8 above)
Comment (tliron) [inline aria/modeling/exceptions.py]: Changing to "Reserved".

File: tests/orchestrator/test_workflow_runner.py (hunk 9 above)
Comment (tliron) [inline tests/orchestrator/test_workflow_runner.py]: No, not in the way this particular test works, since it creates an operation model directly and doesn't call `configure()`

File: aria/modeling/utils.py (hunk 11 above)
Comment (tliron) [inline aria/modeling/utils.py]: The original code was superfluous and complex, doing unpacking and re-packing of values, and actually had a bug in one area of it. I think this code is much cleaner!

File: tests/orchestrator/test_workflow_runner.py (hunk 9 above)
Comment (ran-z) [inline tests/orchestrator/test_workflow_runner.py]: right :+1:

File: aria/modeling/exceptions.py (hunk 8 above)
Comment (ran-z) [inline aria/modeling/exceptions.py]: :+1:

File: extensions/aria_extension_tosca/simple_v1_0/modeling/functions.py (hunk 13 above)
Comment (ran-z) [inline extensions/aria_extension_tosca/simple_v1_0/modeling/functions.py]: :+1:

File: aria/orchestrator/workflows/events_logging.py (hunk 12 above)
Comment (ran-z) [inline aria/orchestrator/workflows/events_logging.py]: :+1:

File: aria/orchestrator/workflow_runner.py (hunk 10 above)
Comment (ran-z) [inline aria/orchestrator/workflow_runner.py]: :+1:

File: aria/modeling/utils.py (hunk 7 above)
Comment (ran-z) [inline aria/modeling/utils.py]: :+1:

File: aria/modeling/utils.py (hunk 11 above)
Comment (ran-z) [inline aria/modeling/utils.py]: :+1:
